src/core/WebGLPathTracer.js

```diff
@@ -428,18 +428,24 @@ export class WebGLPathTracer {
 		const elapsedTime = clock.getElapsedTime() * 1e3;
 		if ( ! this.pausePathTracing && this.enablePathTracing && this.renderDelay <= elapsedTime && ! this.isCompiling ) {
 
-			if ( this.enableDenoiser ) {
+			if ( this.samples < this.maxSamples ) {
 
-				// check to run the denoiser
-				if ( this.isDenoised ) this.denoiser.renderOutput( this.renderAux );
-				else if ( Math.floor( this.samples ) === this.maxSamples ) this.denoiseSample();
+				pathTracer.update();
 
 			}
 
-			// reject if above max samples
-			if ( this.samples >= this.maxSamples ) return;
+			if ( this.enableDenoiser && this.samples === this.maxSamples && this._quad.material.opacity === 1 ) {
 
-			pathTracer.update();
+				if ( ! this.isDenoised && ! this.blockDenoise ) {
+
+					// TODO: if we call this and then perform other draw operations then the result is mangled
+					// so we return here until denoising is complete
+					this.denoiseSample();
+					return;
```

Comment:
If we don't have this return statement and we perform any rendering after `denoiseSample` is called then the image is mangled - likely due to state getting modified on the graphics context. Is it the case that the denoiser is saving, setting, and restoring state before and after performing work?

You should be able to reproduce by removing the return statement here and the `opacity === 1` condition. Here's what it looks like:

<img width="400" alt="image" src="https://github.com/user-attachments/assets/09a54ba6-0c79-47ba-9597-b96b5f940470">